src/components/workspace/editorMode/source.js

```diff
@@ -0,0 +1,831 @@
+                    if (stack[0] === this.nextState) {
+                        stack[1]--;
+                    }
+                    if (!stack[1] || stack[1] < 0) {
+                        stack.splice(0, 2);
+                    }
+                }
+                this.next = stack[0] || "start";
+                return [{type: this.token, value: value}];
+            },
+            regex : "/?>",
+            token : "meta.tag.punctuation.tag-close.xml",
+        },
+        jsxJsRule,
+        comments("jsxAttributes"),
+        {
+            regex : tagRegex,
+            token : "entity.other.attribute-name.xml",
+        }, {
+            regex : "=",
+            token : "keyword.operator.attribute-equals.xml",
+        }, {
+            regex : "\\s+",
+            token : "text.tag-whitespace.xml",
+        }, {
+            push : [
+                {token : "string.attribute-value.xml", regex: "'", next: "pop"},
+                {include : "reference"},
+                {defaultToken : "string.attribute-value.xml"}
+            ],
+            regex : "'",
+            stateName : "jsx_attr_q",
+            token : "string.attribute-value.xml",
+            
+        }, {            
+            push : [
+                {token : "string.attribute-value.xml", regex: '"', next: "pop"},
+                {include : "reference"},
+                {defaultToken : "string.attribute-value.xml"}
+            ],
+            regex : '"',
+            stateName : "jsx_attr_qq",
+            token : "string.attribute-value.xml",
+        },
+        jsxTag
+        ];
+        this.$rules.reference = [{
+            regex : "(?:&#[0-9]+;)|(?:&#x[0-9a-fA-F]+;)|(?:&[a-zA-Z0-9_:\\.-]+;)",
+            token : "constant.language.escape.reference.xml",
+        }];
+    }
+    
+    function comments(next) {
+        return [
+            {
+                next: [
+                    DocCommentHighlightRules.getTagRule(),
+                    {token : "comment", regex : "\\*\\/", next : next || "pop"},
+                    {defaultToken : "comment", caseInsensitive: true}
+                ],
+                regex : /\/\*/,
+                token : "comment", // multi line comment
+                
+            }, {
+                next: [
+                    DocCommentHighlightRules.getTagRule(),
+                    {token : "comment", regex : "$|^", next : next || "pop"},
+                    {defaultToken : "comment", caseInsensitive: true}
+                ],
+                regex : "\\/\\/",
+                token : "comment",
+            }
+        ];
+    }
+    exports.JavaScriptHighlightRules = JavaScriptHighlightRules;
+    });
+    
+    ace.define("ace/mode/matching_brace_outdent",["require","exports","module","ace/range"], function(acequire, exports, module) {
+    "use strict";
+    
+    var Range = acequire("../range").Range;
+    
+    var MatchingBraceOutdent = function() {return};
```

Comment:
What is the purpose of this function? Maybe you could add comments to explain why this is necessary, as it feels quite strange to have an empty function?

Reply:
I deleted the entire ace.define function related to matching brace outdent as I think it is not relevant to the customised source mode.